# Built-in Type Methods - Exercise Solutions

## Instructions

In this workbook, you will dive deeper into Python's string built-in type methods. The exercises will focus on several of the most commonly used string methods. A comprehensive list of all available string methods can be found here: https://www.w3schools.com/python/python_ref_string.asp. As with built-in functions, do not worry about trying to learn or memorize each string method. In the remainder of the course, we will highlight the most useful ones. This workbook will require you to search the provided list of string methods to find the most appropriate one for the situation.

In [9]:
label = "total elevation change"
label = label.capitalize()
print(label)

Total elevation change


In [10]:
name = "sana ur rehman arain"
name = name.title()
print(name)

Sana Ur Rehman Arain


## Intro

As a top analyst for a secret spy agency, you sit under the dim glow of a single desk lamp, surrounded by encrypted messages from a notorious enemy network. Armed with Python string methods, you are well-equipped to decode the hidden messages.

## Exercise 1 - Counting sequences within strings

You've intercepted the following extremely long message.

In [ ]:
code_1 = "Living by the ocean means waking up each day to the rhythmic sound of waves crashing against the shore. The ocean, vast and mesmerizing, becomes a constant backdrop to daily life, its salty breeze a refreshing companion. Every morning, the ocean's expanse invites long, reflective walks along the sandy beach, where the troubles of life seem to dissolve into the ocean's depths. Children play by the ocean, building castles and chasing the retreating waves, their laughter mingling with the ocean's roar. The ocean is also a provider, offering up its bounty to local fishermen whose boats dot the horizon at dawn. As the sun sets, the ocean turns into a canvas of fiery hues, reflecting the sun's last rays in a spectacle of color. At night, the ocean whispers through open windows, lulling residents into a peaceful sleep with its soothing sounds. The ocean's mood can shift, showing a tempestuous side with churning waves and fierce winds that command respect and awe. Through the seasons, the ocean's beauty is a constant, whether it’s under the bright summer sun or the quiet gray of winter. Nearby, the local library, with its windows facing the sea, offers a tranquil retreat where the rhythmic pulse of the ocean continues to soothe those immersed in their books. Living by the ocean is not just a place to reside; it's an endless interaction with the forces of nature, where each day is shaped by the ocean's eternal presence."

You know from previous experience that if the enemy network includes the word `ocean` in their message more than 12 times, they will soon be launching an attack. Your task is to determine how many times `ocean` appears to gauge the urgency of the threat. Use a string method to accomplish this in a single line of code.

**Solution**

In [ ]:
print(f"The word 'ocean' appears in the code {code_1.count('ocean')} times.")


The word 'ocean' appears in the code 14 times.


Oh no! Ocean appears 14 times, the enemy network is looking to launch an attack!

## Exercise 2 - Stripping whitespace and other characters

Soon after you decode the first message, the next coded message arrives and may contain more important information.

In [ ]:
code_2 = "        '''!!Apple Grove    99999        "

Using string methods, remove all the distraction characters (extra spaces, `'`, `!`, `9`).

**Solution**

You could just use the `.strip()` method...

In [ ]:
code_2 = code_2.strip()
code_2 = code_2.strip("'")
code_2 = code_2.strip("!")
code_2 = code_2.strip("9")
code_2 = code_2.strip()
code_2

'Apple Grove'

Or you can be precise and use `.lstrip()` and `.rstrip()` too!

In [ ]:
code_2 = code_2.strip()
code_2 = code_2.lstrip("'")
code_2 = code_2.lstrip("!")
code_2 = code_2.rstrip("9")
code_2 = code_2.rstrip()
code_2

'Apple Grove'

It looks like the enemy may be targeting Apple Grove!

## Exercise 3 - Replacing characters

### 🚨 New Concept: Method chaining

In programming, we often want to apply multiple methods to the same data in sequence. Similar to Exercise 2, we can accomplish this by assigning each intermediate step back to the variable like this...

In [ ]:
my_var = "     text!!!    "
my_var = my_var.strip()
my_var = my_var.rstrip("!")
my_var

'text'

However, Python provides us with a more efficient way to do this called method chaining. Method chaining allows us to apply a method to the result of another method like this...

In [ ]:
my_var = "     text!!!    "
my_var = my_var.strip().rstrip("!")
my_var

'text'

Method chaining works by applying the method to the data from left to right. This means that in this example `.strip()` is applied to `my_var` first, and then `.rstrip()` is applied to the result of `my_var.strip()` second. If we applied the methods in reverse order like this...

In [ ]:
my_var = "     text!!!    "
my_var = my_var.rstrip("!").strip()
my_var

'text!!!'

We get a different result since here `.rstrip()` was applied first.

When chaining lots of methods together, you can use the `\` character at the end of a line to help format your code like this...

In [ ]:
my_var = "     text!!!    "
my_var = my_var.rstrip("!") \
               .strip()
my_var

'text!!!'

### Question

A few minutes later, you receive another message. The most cryptic one yet!

In [ ]:
code_3 = "111000410100010011110010910010100010010101301010100101017000000101011101010111105101010170101010101011101010101100101011181010101010100101001001010010001001010111001010101001010000100100101"

Luckily, you have seen messages like this before so you know how to decode it. You know that the `0` and `1` characters are meaningless distraction characters and you can replace them with the empty string. You also know that the other numbers correspond to the following letters...

- 2 = d
- 3 = b
- 4 = l
- 5 = a
- 6 = m
- 7 = r
- 8 = y
- 9 = i

Use a string method, decode this message on a single line of code.

**Solution**

In [ ]:
result = code_3.replace("0", "") \
               .replace("1","") \
               .replace("2", "d") \
               .replace("3", "b") \
               .replace("4", "l") \
               .replace("5", "a") \
               .replace("6", "m") \
               .replace("7", "r") \
               .replace("8", "y") \
               .replace("9", "i")

print(f"This message decodes to '{result}'.")

This message decodes to 'library'.


Uh oh! Based on the last two messages, it looks like the next target is the library in Apple Grove!

## Exercise 4 - Finding the index of characters or substrings

As you are thinking about how to gather more information on the situation, you recall seeing the word `library` in the original message. You wonder if its position within the message encodes any useful information. Use a string method to determine what position the word `library` occurs at in `code_1`.

**Solution**

In [ ]:
print(f"The word 'library' occurs at character {code_1.find('library')} times.")


The word 'library' occurs at character 1114 times.


Hmmmm... 1114... what could that mean?

## Exercise 5 - Boolean return values

### 🚨 New Concept: Booleans

Say you wanted to know if you can cast a string to an integer. You want to avoid getting an error, so you would like to know first if your string contains only numeric data. Conveniently, the string type provides a method `.isdigit()` that allows us to check whether or not our string contains exclusively numeric values.

In [ ]:
x = "1234"

x.isdigit()

True

As you may have expected, the `.isdigit()` results in `True`. However, this `True` is special. While it looks like a string in the output box, it is actually another Python data type: a boolean.

In [ ]:
type(x.isdigit())

bool

In the next few videos, we will explore how to use booleans and learn why they are useful. For the rest of this workbook, just know that boolean values are either `True` or `False`.

### Question

After some more deep pondering... eureka! 1114 could correspond to the the time of the attack - 11:14 am! Just as you are having this thought, you receive another message...

In [ ]:
code_4 = "10az7"

Initially, you are unsure what this code could mean. But then, you remember that the library has a safe that stores several important historical documents. This message is probably the code for the safe! To confirm this, you want to check that the message is alpha-numeric (meaning it contains only letters and numbers, and not characters like ?!@#$). Use a string method to test whether `code_4` is alpha-numeric.

**Solution**

In [ ]:
code_4.isalnum()

True

It looks like the code is alpha-numeric.

Equipped with all this intelligence you quickly call your boss to inform them that the enemy is planning to attack the Apple Grove library at 11:14 am to break into the safe stored there.

You and the rest of your spy team begin working on counteracting the threat. Great work!

You've intercepted the following extremely long message. You know from previous experience that if the enemy network includes the word ocean in their message more than 12 times, they will soon be launching an attack. Your task is to determine how many times ocean appears to gauge the urgency of the threat. Use a string method to accomplish this in a single line of code.

In [1]:
code_1 = "Living by the ocean means waking up each day to the rhythmic \
    sound of waves crashing against the shore. The ocean, vast and mesmerizing, \
    becomes a constant backdrop to daily life, its salty breeze a refreshing companion. \
    Every morning, the ocean's expanse invites long, reflective walks along the sandy beach, \
    where the troubles of life seem to dissolve into the ocean's depths. Children play by the ocean, \
    building castles and chasing the retreating waves, their laughter mingling with the ocean's roar. \
    The ocean is also a provider, offering up its bounty to local fishermen whose boats dot the horizon at dawn. \
    As the sun sets, the ocean turns into a canvas of fiery hues, reflecting the sun's last rays \
    in a spectacle of color. At night, the ocean whispers through open windows, lulling residents \
    into a peaceful sleep with its soothing sounds. The ocean's mood can shift, showing a tempestuous \
    side with churning waves and fierce winds that command respect and awe. Through the seasons, \
    the ocean's beauty is a constant, whether it’s under the bright summer sun or the quiet gray of winter. \
    Nearby, the local library, with its windows facing the sea, offers a tranquil retreat where the rhythmic \
    pulse of the ocean continues to soothe those immersed in their books. Living by the ocean is not just a place \
    to reside; it's an endless interaction with the forces of nature, where each day is shaped by the ocean's eternal presence." # you can use the \ character at the end of a line to help format your code like this...

print(f"The word 'ocean' appears in the code {code_1.count('ocean')} times.")

The word 'ocean' appears in the code 14 times.


Using string methods, remove all the distraction characters (extra spaces, ', !, 9).

In [2]:
code_2 = "        '''!!Apple Grove    99999        "
cleaned_code_2 = code_2.strip(" '9!") # Removes leading and trailing characters specified in the argument 
print(cleaned_code_2)  

Apple Grove


In programming, we often want to apply multiple methods to the same data in sequence. we can accomplish this by assigning each intermediate step back to the variable like this...

In [3]:
my_var = "     text!!!    "
my_var = my_var.strip()
my_var = my_var.rstrip("!")

print(my_var)

text


However, Python provides us with a more efficient way to do this called method chaining. Method chaining allows us to apply a method to the result of another method like this...

In [4]:
my_var_1 = "     sentence!!!    "
my_var_1 = my_var_1.strip().rstrip("!")
print(my_var_1)

sentence


In [5]:
# you receive another message. The most cryptic one yet! Luckily, 
# you have seen messages like this before so you know how to decode it. 
# You know that the 0 and 1 characters are meaningless distraction characters 
# and you can replace them with the empty string. You also know that the other numbers correspond to the following letters...
# 2 = d
# 3 = b
# 4 = l
# 5 = a
# 6 = m 
# 7 = r
# 8 = y
# 9 = i 
# Use a string method, decode this message on a single line of code.

code_3 = "11100041010001001111001091001010001001010130101010010101700000010101110101011110510101017010101\
    0101011101010101100101011181010101010100101001001010010001001010111001010101001010000100100101"

decoded_code = code_3.replace("0", "")\
                    .replace("1", "")\
                    .replace("2", "d")\
                    .replace("3", "b")\
                    .replace("4", "l")\
                    .replace("5", "a")\
                    .replace("6", "m")\
                    .replace("7", "r")\
                    .replace("8", "y")\
                    .replace("9", "i")

print(f"This message decodes to '{decoded_code}'.")

This message decodes to 'librar    y'.


As you are thinking about how to gather more information on the situation, you recall seeing the word library in the original message. You wonder if its position within the message encodes any useful information. Use a string method to determine what position the word library occurs at in code_1.

In [6]:
position_library = code_1.find("library")
print(f"The word 'library' occures at character {position_library}.")

The word 'library' occures at character 1162.


Say you wanted to know if you can cast a string to an integer. You want to avoid getting an error, so you would like to know first if your string contains only numeric data. Conveniently, the string type provides a method .isdigit() that allows us to check whether or not our string contains exclusively numeric values.

In [7]:
x = "1234"
print(x.isdigit())

True


You want to check that the message is alpha-numeric (meaning it contains only letters and numbers, and not characters like ?!@#$). Use a string method to test whether code_4 is alpha-numeric.

In [8]:
code_4 = "10az7"

print(code_4.isalnum())

True
